### PR TITLE
`server` refactor

### PR DESCRIPTION
**Pull request by danlaine**
Significantly changes the implementation of `server/server.go` and `server/network.go`

* Add lock to `server` and `localNetwork` 
* Remove use of channels that I found hard to follow
* Miscellaneous cleanup of names, declarations, etc. 

Still need to do another pass over this and do more testing.

### CODE REVIEW

File: server/network.go [hunk 1]
```diff
@@ -358,97 +324,119 @@ func (lc *localNetwork) loadSnapshot(
...
 
-	// node info is already available
 	if err := lc.updateNodeInfo(); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (lc *localNetwork) loadSnapshotWait(ctx context.Context, loadSnapshotReadyCh chan struct{}) {
```
Comment (danlaine) [inline server/network.go:369]: replaced by `AwaitHealthy`

File: server/server.go [hunk 2]
```diff
@@ -27,14 +27,41 @@ import (
...
 	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
 	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/credentials/insecure"
 	"google.golang.org/grpc/status"
 )
 
+const (
```
Comment (danlaine) [inline server/server.go:40]: moved up from below

File: server/network.go [hunk 3]
```diff
@@ -358,97 +339,127 @@ func (lc *localNetwork) loadSnapshot(
...
-		nodeInfo := lc.nodeInfos[name]
+	if err := lc.updateNodeInfo(); err != nil {
+		return err
+	}
+
+	if err := lc.updateSubnetInfo(ctx); err != nil {
+		return err
+	}
+
+	// TODO do we need to log this? Seems verbose
```
Comment (danlaine) [inline server/network.go:432]: Open question for you guys

File: server/server.go [hunk 4]
```diff
@@ -404,136 +380,48 @@ func (s *server) waitChAndUpdateClusterInfo(msg string, readyCh chan struct{}, u
...
-			continueLoop = false
-			err = ctx.Err()
+			return &rpcpb.WaitForHealthyResponse{ClusterInfo: s.clusterInfo}, ctx.Err()
 		case <-time.After(1 * time.Second):
 		}
 	}
-	return &rpcpb.WaitForHealthyResponse{ClusterInfo: s.getClusterInfo()}, err
-}
-
-func getNetworkBlockchainSpec(
```
Comment (danlaine) [inline server/server.go:430]: moved to bottom of file

File: server/network.go (hunk 3 above)
Comment (holisticode) [reply server/network.go:432]: Possibly at most Debug log

File: server/server.go [hunk 5]
```diff
@@ -1022,92 +904,98 @@ func (s *server) LoadSnapshot(ctx context.Context, req *rpcpb.LoadSnapshotReques
...
+		defer cancel()
+		err := s.network.AwaitHealthy(ctx)
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		if err != nil {
+			s.log.Warn("snapshot load failed to complete. stopping network and cleaning up network", zap.Error(err))
+
+			s.stopAndRemoveNetwork()
+			return
+		} else {
```
Comment (holisticode) [inline server/server.go:935]: Maybe we don't need an else here as we are return in the != nil case?

File: server/network.go (hunk 3 above)
Comment (danlaine) [reply server/network.go:432]: I'll leave it to you guys to decide -- feel free to push a commit changing this as you see fit

File: cmd/server/server.go [hunk 6]
```diff
@@ -82,31 +85,37 @@ func serverFunc(*cobra.Command, []string) (err error) {
...
-	rootCtx, rootCancel := context.WithCancel(context.Background())
-	errc := make(chan error)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	errChan := make(chan error)
 	go func() {
-		errc <- s.Run(rootCtx)
+		err := s.Run(ctx)
+		if err != nil {
+			log.Error("server Run error", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (danlaine) [reply cmd/server/server.go:100]: Up to you -- feel free to push a commit changing this

File: cmd/server/server.go [hunk 7]
```diff
@@ -82,31 +85,37 @@ func serverFunc(*cobra.Command, []string) (err error) {
...
 		log.Warn("closed server", zap.Error(waitForServerStop))
-	case serverClosed := <-errc:
-		// server already stopped here
-		_ = rootCancel
+	case serverClosed := <-errChan:
+		// The server stopped.
 		log.Warn("server closed", zap.Error(serverClosed))
 	}
-	return err
+	return nil
```
[the remark this answers is not in the document]
Comment (danlaine) [reply cmd/server/server.go:120]: This is logged in the above `select`

File: server/server.go (hunk 5 above)
Comment (danlaine) [reply server/server.go:934]: fixed

File: server/network.go [hunk 8]
```diff
@@ -195,7 +184,16 @@ func (lc *localNetwork) createConfig() error {
...
 }
 
-func (lc *localNetwork) start() error {
+// Creates a network and sets [lc.nw] to it.
+// Assumes [lc.lock] isn't held.
+func (lc *localNetwork) Start() error {
+	lc.lock.Lock()
+	defer lc.lock.Unlock()
+
+	if lc.nw != nil {
```
[the remark this answers is not in the document]
Comment (danlaine) [reply server/network.go:193]: Same comment as https://github.com/ava-labs/avalanche-network-runner/pull/447#discussion_r1119112807

Would you mind elaborating a bit?

File: server/server.go [hunk 9]
```diff
@@ -27,14 +27,39 @@ import (
...
+	waitForHealthyTimeout = 2 * time.Minute
+
+	rootDataDirPrefix = "network-runner-root-data"
+)
+
+var (
+	ErrInvalidVMName          = errors.New("invalid VM name")
+	ErrInvalidPort            = errors.New("invalid port")
+	ErrNotEnoughNodesForStart = errors.New("not enough nodes specified for start")
+	ErrNetworkRunning         = errors.New("network already running")
```
Comment (felipemadero) [inline server/server.go:55]: I don't think we need to change Err names on err description. This errors where already avaiable with different naming. And they are also part of interface with clients.

File: server/server.go [hunk 10]
```diff
@@ -232,29 +238,28 @@ func (s *server) Ping(context.Context, *rpcpb.PingRequest) (*rpcpb.PingResponse,
...
-	if deadline, ok := ctx.Deadline(); !ok || time.Until(deadline) < defaultStartTimeout {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(context.Background(), defaultStartTimeout)
-		_ = cancel // don't call since "start" is async, "curl" may not specify timeout
-		s.log.Info("received start request with default timeout", zap.String("timeout", defaultStartTimeout.String()))
-	} else {
-		s.log.Info("received start request with existing timeout", zap.String("timeout", deadline.String()))
+	// If [network] is already populated, the network has already been started.
+	if s.network != nil {
+		return nil, ErrNetworkRunning
```
Comment (felipemadero) [inline server/server.go:247]: can use old naming? applies other places also

File: server/server.go [hunk 11]
```diff
@@ -335,67 +321,58 @@ func (s *server) Start(ctx context.Context, req *rpcpb.StartRequest) (*rpcpb.Sta
...
-	readyCh := make(chan struct{})
-	go s.network.startWait(ctx, chainSpecs, readyCh)
-
 	// update cluster info non-blocking
 	// the user is expected to poll this latest information
 	// to decide cluster/subnet readiness
 	go func() {
-		s.waitChAndUpdateClusterInfo("local cluster", readyCh, false)
-		s.waitChAndUpdateClusterInfo("custom chains", readyCh, true)
+		ctx, cancel := context.WithTimeout(context.Background(), waitForHealthyTimeout)
```
Comment (felipemadero) [inline server/server.go:348]: previous semantics was to update info just after network start (before creating chains), and after creating chains. so, almost instantaneously the user can check cluster and node info, without waiting for longer chain creation operation for that

File: server/server.go [hunk 12]
```diff
@@ -232,29 +238,28 @@ func (s *server) Ping(context.Context, *rpcpb.PingRequest) (*rpcpb.PingResponse,
 	return &rpcpb.PingResponse{Pid: int32(os.Getpid())}, nil
 }
 
-const defaultStartTimeout = 5 * time.Minute
+func (s *server) Start(_ context.Context, req *rpcpb.StartRequest) (*rpcpb.StartResponse, error) {
+	s.mu.Lock()
```
Comment (felipemadero) [inline server/server.go:242]: this is a lock over server network/clusterinfo data, another PR was trying to reduce the space of server mutual exclusion by reducing the lock only to parts of code that modify them. This PR is expanding the lock even more

File: server/server.go [hunk 13]
```diff
@@ -335,67 +321,58 @@ func (s *server) Start(ctx context.Context, req *rpcpb.StartRequest) (*rpcpb.Sta
...
-				s.clusterInfo.CustomChains[chainID.String()] = chainInfo.info
-			}
-			s.clusterInfo.Subnets = s.network.subnets
-		}
-		s.log.Info(fmt.Sprintf("%s ready", msg))
-		s.mu.Unlock()
+// Asssumes [s.mu] is held.
+func (s *server) updateClusterInfo() {
+	s.clusterInfo.Healthy = true
+	s.clusterInfo.NodeNames = maps.Keys(s.network.nodeInfos)
```
Comment (felipemadero) [inline server/server.go:368]: this is unsortered

File: server/server.go [hunk 14]
```diff
@@ -404,136 +381,48 @@ func (s *server) waitChAndUpdateClusterInfo(msg string, readyCh chan struct{}, u
...
+		s.mu.RLock()
+
+		if s.network == nil {
+			s.mu.RUnlock()
+			return nil, ErrNetworkNotRunning
+		}
 		if s.clusterInfo.CustomChainsHealthy {
-			break
+			s.mu.RUnlock()
+			return &rpcpb.WaitForHealthyResponse{ClusterInfo: s.clusterInfo}, nil
```
Comment (felipemadero) [inline server/server.go:397]: no lock to access this?

File: server/server.go [hunk 15]
```diff
@@ -404,136 +381,48 @@ func (s *server) waitChAndUpdateClusterInfo(msg string, readyCh chan struct{}, u
...
 		}
+		s.mu.RUnlock()
+
 		select {
-		case err = <-s.asyncErrCh:
-			continueLoop = false
 		case <-ctx.Done():
-			continueLoop = false
-			err = ctx.Err()
+			return &rpcpb.WaitForHealthyResponse{ClusterInfo: s.clusterInfo}, ctx.Err()
```
Comment (felipemadero) [inline server/server.go:403]: same here, no lock?

File: server/server.go [hunk 16]
```diff
@@ -543,143 +432,152 @@ func (s *server) CreateBlockchains(
...
-	if err := s.getNetwork().waitForLocalClusterReady(ctx); err != nil {
+	if err := s.network.AwaitHealthy(ctx); err != nil {
 		return nil, err
 	}
 
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	s.clusterInfo.NodeNames = s.network.nodeNames
+	s.clusterInfo.NodeNames = maps.Keys(s.network.nodeInfos)
```
Comment (felipemadero) [inline server/server.go:528]: unsorted

File: server/server.go [hunk 17]
```diff
@@ -543,143 +432,152 @@ func (s *server) CreateBlockchains(
...
-	go s.network.createSubnets(ctx, numSubnets, readyCh)
-
 	// update cluster info non-blocking
 	// the user is expected to poll this latest information
 	// to decide cluster/subnet readiness
 	go func() {
-		s.waitChAndUpdateClusterInfo("custom chains", readyCh, true)
+		ctx, cancel := context.WithTimeout(context.Background(), waitForHealthyTimeout)
+		defer cancel()
+		err := s.network.CreateSubnets(ctx, numSubnets)
```
Comment (felipemadero) [inline server/server.go:496]: we want this error to be returned by WaitForHealthy

File: server/server.go [hunk 18]
```diff
@@ -335,67 +321,58 @@ func (s *server) Start(ctx context.Context, req *rpcpb.StartRequest) (*rpcpb.Sta
...
-
 	// update cluster info non-blocking
 	// the user is expected to poll this latest information
 	// to decide cluster/subnet readiness
 	go func() {
-		s.waitChAndUpdateClusterInfo("local cluster", readyCh, false)
-		s.waitChAndUpdateClusterInfo("custom chains", readyCh, true)
+		ctx, cancel := context.WithTimeout(context.Background(), waitForHealthyTimeout)
+		defer cancel()
+		err := s.network.CreateChains(ctx, chainSpecs)
```
Comment (felipemadero) [inline server/server.go:350]: we want this error to be returned by WaitForHealthy

File: server/server.go [hunk 19]
```diff
@@ -404,136 +381,48 @@ func (s *server) waitChAndUpdateClusterInfo(msg string, readyCh chan struct{}, u
...
+		}
 		if s.clusterInfo.CustomChainsHealthy {
-			break
+			s.mu.RUnlock()
+			return &rpcpb.WaitForHealthyResponse{ClusterInfo: s.clusterInfo}, nil
 		}
+		s.mu.RUnlock()
+
 		select {
-		case err = <-s.asyncErrCh:
```
Comment (felipemadero) [inline server/server.go:419]: we want to notify the cause of async error here (cause for the network to fail)

File: server/server.go [hunk 20]
```diff
@@ -543,143 +432,152 @@ func (s *server) CreateBlockchains(
...
-	go s.network.createBlockchains(ctx, chainSpecs, readyCh)
-
 	// update cluster info non-blocking
 	// the user is expected to poll this latest information
 	// to decide cluster/subnet readiness
 	go func() {
-		s.waitChAndUpdateClusterInfo("custom chains", readyCh, true)
+		ctx, cancel := context.WithTimeout(context.Background(), waitForHealthyTimeout)
+		defer cancel()
+		err := s.network.CreateChains(ctx, chainSpecs)
```
Comment (felipemadero) [inline server/server.go:454]: we want this error to be returned by WaitForHealthy

File: server/server.go [hunk 21]
```diff
@@ -812,50 +712,50 @@ func (s *server) AddNode(_ context.Context, req *rpcpb.AddNodeRequest) (*rpcpb.A
...
 		return nil, err
 	}
 
-	if err := s.network.updateNodeInfo(); err != nil {
+	if err := s.network.UpdateNodeInfo(); err != nil {
 		return nil, err
 	}
 
-	s.clusterInfo.NodeNames = s.network.nodeNames
+	s.clusterInfo.NodeNames = maps.Keys(s.network.nodeInfos)
```
Comment (felipemadero) [inline server/server.go:743]: unsorted

File: server/server.go [hunk 22]
```diff
@@ -869,36 +769,25 @@ func (s *server) RestartNode(ctx context.Context, req *rpcpb.RestartNodeRequest)
...
```
Comment (felipemadero) [inline server/server.go:776]: same

File: server/server.go [hunk 23]
```diff
@@ -869,36 +769,25 @@ func (s *server) RestartNode(ctx context.Context, req *rpcpb.RestartNodeRequest)
...
 
-	info := s.clusterInfo
-	if info == nil {
-		info = &rpcpb.ClusterInfo{}
-	}
-	info.Healthy = false
+	s.stopAndRemoveNetwork()
 
-	return &rpcpb.StopResponse{ClusterInfo: info}, nil
+	return &rpcpb.StopResponse{}, nil
```
Comment (felipemadero) [inline server/server.go:790]: previously this was returning latest valid cluster info

File: server/server.go [hunk 24]
```diff
@@ -1022,92 +903,98 @@ func (s *server) LoadSnapshot(ctx context.Context, req *rpcpb.LoadSnapshotReques
...
-	go s.network.loadSnapshotWait(ctx, readyCh)
-
 	// update cluster info non-blocking
 	// the user is expected to poll this latest information
 	// to decide cluster/subnet readiness
 	go func() {
-		s.waitChAndUpdateClusterInfo("local cluster", readyCh, true)
+		ctx, cancel := context.WithTimeout(context.Background(), waitForHealthyTimeout)
+		defer cancel()
+		err := s.network.AwaitHealthy(ctx)
```
Comment (felipemadero) [inline server/server.go:926]: we want this error to be returned by WaitForHealthy

File: rpcpb/rpc.proto [hunk 25]
```diff
@@ -407,9 +407,7 @@ message AddNodeResponse {
 
 message StopRequest {}
 
-message StopResponse {
```
[the remark this answers is not in the document]
Comment (danlaine) [reply rpcpb/rpc.proto:410]: The response is always an empty cluster info because the network is stopped

File: tests/e2e/e2e_test.go [hunk 26]
```diff
@@ -627,7 +627,7 @@ var _ = ginkgo.Describe("[Start/Remove/Restart/Add/Stop]", func() {
 			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
 			_, err := cli.Health(ctx)
 			cancel()
-			gomega.Ω(err.Error()).Should(gomega.ContainSubstring("not bootstrapped"))
+			gomega.Ω(err.Error()).Should(gomega.ContainSubstring(server.ErrNetworkNotRunning.Error()))
```
[the remark this answers is not in the document]
Comment (danlaine) [reply tests/e2e/e2e_test.go:630]: It wasn't accurate -- it was saying "not bootstrapped" but really the error is that the network isn't running at all

File: tests/e2e/e2e_test.go [hunk 27]
```diff
@@ -496,7 +496,7 @@ var _ = ginkgo.Describe("[Start/Remove/Restart/Add/Stop]", func() {
 		})
 		ginkgo.By("calling start API with custom config", func() {
 			ux.Print(log, logging.Green.Wrap("sending 'start' with the valid binary path: %s"), execPath1)
-			ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
```
[the remark this answers is not in the document]
Comment (danlaine) [reply tests/e2e/e2e_test.go:499]: It was flaking so I bumped it

File: cmd/server/server.go [hunk 28]
```diff
@@ -60,16 +60,19 @@ func serverFunc(*cobra.Command, []string) (err error) {
...
 	}
-	lcfg := logging.Config{
-		DisplayLevel: lvl,
-		LogLevel:     lvl,
-	}
-	lcfg.Directory = logDir
-	logFactory := logging.NewFactory(lcfg)
+
+	logFactory := logging.NewFactory(logging.Config{
+		RotatingWriterConfig: logging.RotatingWriterConfig{
```
[the remark this answers is not in the document]
Comment (danlaine) [reply cmd/server/server.go:70]: This is semantically the same as what was here before

File: cmd/server/server.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (danlaine) [reply cmd/server/server.go:100]: I'll defer to you guys on logging -- whatever you want

File: server/network.go [hunk 29]
```diff
@@ -20,10 +19,12 @@ import (
 	"github.com/ava-labs/avalanchego/ids"
 	avago_constants "github.com/ava-labs/avalanchego/utils/constants"
 	"github.com/ava-labs/avalanchego/utils/logging"
+	"golang.org/x/exp/maps"
 )
 
 type localNetwork struct {
-	log logging.Logger
+	lock sync.Mutex
```
[the remark this answers is not in the document]
Comment (danlaine) [reply server/network.go:26]: We start `localNetwork` methods in a goroutine in some places in `server`, so we need a lock to ensure we're not concurrently accessing the state of `localNetwork`

File: server/network.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (danlaine) [reply server/network.go:193]: This check might not be _strictly_ necessary but it makes the code pattern nicer -- namely, exported methods use a lock

File: server/network.go [hunk 30]
```diff
@@ -215,122 +213,105 @@ func (lc *localNetwork) start() error {
...
-	var ctx context.Context
-	ctx, lc.startCtxCancel = context.WithCancel(argCtx)
+) error {
+	lc.lock.Lock()
+	defer lc.lock.Unlock()
 
-	if err := lc.waitForLocalClusterReady(ctx); err != nil {
-		lc.startErrCh <- err
-		return
+	if lc.nw == nil {
```
[the remark this answers is not in the document]
Comment (danlaine) [reply server/network.go:225]: https://github.com/ava-labs/avalanche-network-runner/pull/447#discussion_r1120171717

^ Same as this comment

File: server/network.go [hunk 31]
```diff
@@ -215,122 +213,105 @@ func (lc *localNetwork) start() error {
...
+			cancel()
+		case <-ctx.Done():
+			// This method is done. Don't leak [ctx].
+		}
+	}(ctx)
 
-	if err := lc.updateNodeInfo(); err != nil {
-		lc.startErrCh <- err
-		return
+	if err := lc.awaitHealthy(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (danlaine) [reply server/network.go:246]: I'll leave the naming up to you all

File: server/network.go [hunk 32]
```diff
@@ -195,7 +184,16 @@ func (lc *localNetwork) createConfig() error {
 	return nil
 }
 
-func (lc *localNetwork) start() error {
+// Creates a network and sets [lc.nw] to it.
+// Assumes [lc.lock] isn't held.
+func (lc *localNetwork) Start() error {
```
[the remark this answers is not in the document]
Comment (danlaine) [reply server/network.go:189]: Exported (capitalized) methods are supposed to be used by external callers and grab a lock, non-exported methods aren't and don't grab a lock.

File: server/network.go [hunk 33]
```diff
@@ -358,97 +339,127 @@ func (lc *localNetwork) loadSnapshot(
...
-}
-
+// Populates [lc.customChainIDToInfo] for all chains other than those on
+// the Primary Network (P-Chain, X-Chain, C-Chain.)
+// Populates [lc.subnets] with all subnets that exist.
+// Doesn't contain the Primary network.
+// Assumes [lc.lock] is held.
 func (lc *localNetwork) updateSubnetInfo(ctx context.Context) error {
-	node, err := lc.nw.GetNode(lc.nodeNames[0])
+	nodeNames := maps.Keys(lc.nodeInfos)
```
[the remark this answers is not in the document]
Comment (danlaine) [reply server/network.go:355]: Does it matter which node we query this info from?

File: server/server.go (hunk 9 above)
Comment (danlaine) [reply server/server.go:55]: I don't think we had errors for "network already running" or "no network is running"

File: server/server.go (hunk 10 above)
Comment (danlaine) [reply server/server.go:247]: The old naming was misleading -- it said "not bootstrapped" but it really meant "not running"

File: server/server.go (hunk 11 above)
Comment (danlaine) [reply server/server.go:348]: We update `clusterInfo` before starting the network:

```go
	s.clusterInfo = &rpcpb.ClusterInfo{
		Pid:         pid,
		RootDataDir: rootDataDir,
	}
```

And also after:

```go
	go func() {
		ctx, cancel := context.WithTimeout(context.Background(), waitForHealthyTimeout)
		defer cancel()
		err := s.network.CreateChains(ctx, chainSpecs)
		s.mu.Lock()
		defer s.mu.Unlock()
		if err != nil {
			s.log.Error("network never became healthy", zap.Error(err))
			s.stopAndRemoveNetwork()
			return
		}
		s.updateClusterInfo()
		s.log.Info("network healthy")
	}()
```

File: server/server.go (hunk 12 above)
Comment (danlaine) [reply server/server.go:242]: > This PR is expanding the lock even more

I think this is what it takes in order for `server` to be threadsafe -- we can't have multiple goroutines simultaneously accessing the state of `server` or else it's unsafe

File: server/server.go (hunk 18 above)
Comment (danlaine) [reply server/server.go:350]: https://github.com/ava-labs/avalanche-network-runner/pull/447#discussion_r1120195954

File: server/server.go (hunk 19 above)
Comment (danlaine) [reply server/server.go:419]: I don't understand -- this channel doesn't exist anymore. Where are the unhandled errors?

File: server/server.go (hunk 20 above)
Comment (danlaine) [reply server/server.go:454]: https://github.com/ava-labs/avalanche-network-runner/pull/447#discussion_r1120195954

File: server/server.go (hunk 23 above)
Comment (danlaine) [reply server/server.go:790]: The clusterInfo will always be empty so I don't see a need to return it

File: server/server.go [hunk 34]
```diff
@@ -968,45 +866,28 @@ func (s *server) SendOutboundMessage(ctx context.Context, req *rpcpb.SendOutboun
...
-	} else {
-		s.log.Info("received start request with existing timeout", zap.String("timeout", deadline.String()))
-	}
-
-	// If [network] is already populated, the network has already been started.
-	if s.getNetwork() != nil {
-		return nil, ErrAlreadyBootstrapped
-	}
-
+func (s *server) LoadSnapshot(_ context.Context, req *rpcpb.LoadSnapshotRequest) (*rpcpb.LoadSnapshotResponse, error) {
```
[the remark this answers is not in the document]
Comment (danlaine) [reply server/server.go:869]: It isn't used in this method -- this method is non-blocking

File: server/server.go (hunk 24 above)
Comment (danlaine) [reply server/server.go:926]: Not sure what you mean

File: server/server.go (hunk 21 above)
Comment (danlaine) [reply server/server.go:749]: sorted now

File: server/server.go (hunk 16 above)
Comment (danlaine) [reply server/server.go:532]: sorted now

File: server/server.go (hunk 15 above)
Comment (danlaine) [reply server/server.go:407]: fixed

File: server/server.go (hunk 14 above)
Comment (danlaine) [reply server/server.go:398]: fixed

File: server/server.go (hunk 13 above)
Comment (danlaine) [reply server/server.go:368]: sorted now

File: server/server.go (hunk 10 above)
Comment (danlaine) [reply server/server.go:247]: https://github.com/ava-labs/avalanche-network-runner/pull/447#discussion_r1120182965

File: server/network.go [hunk 35]
```diff
@@ -358,97 +339,127 @@ func (lc *localNetwork) loadSnapshot(
...
+	if err := lc.updateNodeInfo(); err != nil {
+		return err
+	}
+
+	if err := lc.updateSubnetInfo(ctx); err != nil {
+		return err
+	}
+
+	// TODO do we need to log this? Seems verbose
+	for name, nodeInfo := range lc.nodeInfos {
```
[the remark this answers is not in the document]
Comment (danlaine) [reply server/network.go]: Sorted now

File: server/network.go [hunk 36]
```diff
@@ -358,97 +339,127 @@ func (lc *localNetwork) loadSnapshot(
...
 	lc.subnets = []string{}
 	for _, subnet := range subnets {
 		if subnet.ID != avago_constants.PlatformChainID {
 			lc.subnets = append(lc.subnets, subnet.ID.String())
 		}
 	}
-	for _, nodeName := range lc.nodeNames {
-		nodeInfo := lc.nodeInfos[nodeName]
+
+	for _, nodeInfo := range lc.nodeInfos {
```
[the remark this answers is not in the document]
Comment (danlaine) [reply server/network.go]: sorted now

File: cmd/server/server.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/server/server.go:100]: pushing a commit for changing this

File: rpcpb/rpc.proto (hunk 25 above)
[the remark this answers is not in the document]
Comment (felipemadero) [reply rpcpb/rpc.proto:410]: I sugest you to check the output of the command `go run main.go control stop`

[the remark this answers is not in the document]
Comment (felipemadero) [reply rpcpb/rpc.proto:410]: I would like to preserve actual ANR behaviour whenever posible

File: tests/e2e/e2e_test.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (felipemadero) [reply tests/e2e/e2e_test.go:630]: Could we keep the original naming? This will propagate into CLI and represents and API change, would like to include it in a bigger refactor related to API changes with appropiate semver change.

File: server/network.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (felipemadero) [reply server/network.go:194]: let's just remove the check

File: server/network.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (felipemadero) [reply server/network.go:226]: let's remove this

File: server/network.go (hunk 33 above)
[the remark this answers is not in the document]
Comment (felipemadero) [reply server/network.go:356]: Yup it is best for debug to know which node is interacting ANR with

File: server/server.go (hunk 9 above)
Comment (felipemadero) [reply server/server.go:55]: we had ErrAlreadyBootstrapped and ErrNotBootstrapped with same semantics. Would prefer to keep them until bigger change to API.

File: server/server.go (hunk 11 above)
Comment (felipemadero) [reply server/server.go:348]: the previous situation was for clusterInfo to be updated after starting the network, with the node names etc, and secondly after creating the blockchains, with the subnet/blockchain infoi

Comment (felipemadero) [reply server/server.go:348]: anyway we can keep this, it is a minor semantic change not noticeable for users

File: server/server.go (hunk 17 above)
Comment (felipemadero) [reply server/server.go:500]: WaitForHealthy not only waits for network to be on, it is also giving to the client the reason for the network failure if there was any. Usually that error was hidden in ANR logs, recently we added this function so users like CLI can check the error and provide it to the user.
All the async errors from go routines, due to calling CreateChains,  CreateSubnets, AwaitAndUpdateNetworkInfo, should be passed to this function using some mechanism.

File: server/server.go (hunk 19 above)
Comment (felipemadero) [reply server/server.go:419]: https://github.com/ava-labs/avalanche-network-runner/pull/447#discussion_r1131592992

File: server/server.go (hunk 23 above)
Comment (felipemadero) [reply server/server.go:798]: clusterinfo was being returned on stop, pls try to keep the same semantics

File: server/server.go (hunk 34 above)
[the remark this answers is not in the document]
Comment (felipemadero) [reply server/server.go:877]: ok... I understood that maybe the ANR client will want to provide a timeout on the async op. is that reasonable?

File: server/server.go (hunk 24 above)
Comment (felipemadero) [reply server/server.go:934]: https://github.com/ava-labs/avalanche-network-runner/pull/447#discussion_r1131592992

File: server/server.go (hunk 34 above)
[the remark this answers is not in the document]
Comment (danlaine) [reply server/server.go:877]: We can't use the context passed into `LoadSnapshot` for the async operations in the goroutine started by this method because the `ctx` will probably be cancelled after `LoadSnapshot` returns, which will kill the async goroutine.

Namely, the caller is probably doing something like:

```go
ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
result, err := server.LoadSnapshot(ctx, req)
cancel() // This kills the async goroutine even though we didn't mean to
```

**Review by dasconnor — APPROVED** [review]
LGTM, great job everyone!
